flink-libraries/flink-ml/src/main/scala/org/apache/flink/ml/outlier/StochasticOutlierSelection.scala

```diff
@@ -0,0 +1,367 @@
+    *         probabilities, which is in fact the affinity where each row sums up to one.
+    */
+  def computeBindingProbabilities(affinityVectors: DataSet[BreezeLabeledVector]):
+  DataSet[BreezeLabeledVector] =
+  affinityVectors.map(vec => BreezeLabeledVector(vec.idx, vec.data :/ sum(vec.data)))
+
+  /** Compute the final outlier probability by taking the product of the column.
+    *
+    * @param bindingProbabilityVectors The binding probability vectors where the binding
+    *                                  probability is based on the affinity and represents the
+    *                                  probability of a vector binding with another vector.
+    * @return Returns a single double which represents the final outlierness of the input vector.
+    */
+  def computeOutlierProbability(bindingProbabilityVectors: DataSet[BreezeLabeledVector]):
+  DataSet[(Int, Double)] = bindingProbabilityVectors
+    .flatMap(vec => vec.data.toArray.zipWithIndex.map(pair => {
+
+      // The DistanceMatrix removed the diagonal, but we need to compute the product
+      // of the column, so we need to correct the offset.
+      val columnIndex = if (pair._2 >= vec.idx) {
+        1
+      } else {
+        0
+      }
+
+      (columnIndex + pair._2, pair._1)
+    })).groupBy(0).reduceGroup {
+    probabilities => {
+      var rowNumber = -1
+      var outlierProbability = 1.0
+      for (probability <- probabilities) {
+        rowNumber = probability._1
+        outlierProbability = outlierProbability * (1.0 - probability._2)
+      }
+
+      (rowNumber, outlierProbability)
+    }
+  }
+
+  /** Performs a binary search to get affinities in such a way that each conditional Gaussian has
+    *  the same perplexity.
+    *
+    * @param dissimilarityVector The input dissimilarity vector which represents the current
+    *                            vector distance to the other vectors in the data set
+    * @param logPerplexity The log of the perplexity, which represents the probability of having
+    *                      affinity with another vector.
+    * @param maxIterations The maximum iterations to limit the computational time.
+    * @param tolerance The allowed tolerance to sacrifice precision for decreased computational
+    *                  time.
+    * @param beta: The current beta
+    * @param betaMin The lower bound of beta
+    * @param betaMax The upper bound of beta
+    * @param iteration The current iteration
+    * @return Returns the affinity vector of the input vector.
+    */
+  def binarySearch(dissimilarityVector: BreezeVector[Double],
+                   logPerplexity: Double,
+                   maxIterations: Int,
+                   tolerance: Double,
+                   beta: Double = 1.0,
+                   betaMin: Double = Double.NegativeInfinity,
+                   betaMax: Double = Double.PositiveInfinity,
+                   iteration: Int = 0): BreezeVector[Double] = {
```

Comment:
The non written down coding style was to break a long parameter list by writing each parameter in its own line with double indentation. Something like
```
def fun(
    x: Int,
    y: Double)
  : String = {
  // foobar
}